pageserver/src/layered_repository.rs

```diff
@@ -1944,41 +1938,71 @@ impl LayeredTimeline {
         Ok(new_path)
     }
 
+    ///
+    /// Collect a bunch of Level 0 layer files, and compact and reshuffle them as
+    /// as Level 1 files.
+    ///
     fn compact_level0(&self, target_file_size: u64) -> Result<()> {
         let layers = self.layers.read().unwrap();
-
-        let level0_deltas = layers.get_level0_deltas()?;
+        let mut level0_deltas = layers.get_level0_deltas()?;
+        drop(layers);
 
         // We compact or "shuffle" the level-0 delta layers when they've
         // accumulated over the compaction threshold.
-        if level0_deltas.len() < self.get_compaction_threshold() {
+        if level0_deltas.is_empty() || level0_deltas.len() < self.get_compaction_threshold() {
```

Comment:
May be it is better to prohibit zero value for compaction threshold or treat it as "infinity" (prohibit compaction)?